### PR TITLE
[Do not merge] PoC: Add new APIs to handle deep links and process RCBilling redemption links

### PR DESCRIPTION
**Pull request by tonidero**
### Description
This is a PoC of what it would take to listen to deep links to be able to redeem RCBilling purchases in the app using custom schema urls and using an old aliasing endpoint. Note that it has quite a few things that are not final and need to change before merging.

This PoC includes 2 new APIs:
- A `DeepLinkHandler.handleDeepLink(intent)` to input deep links by the app
- A `RedeemWebPurchaseListener` to accept and process web redemption links

### CODE REVIEW

File: examples/purchase-tester/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -15,12 +15,19 @@
             android:theme="@style/AppTheme">
         <activity
                 android:name="com.revenuecat.purchasetester.MainActivity"
+                android:launchMode="singleTop"
```
Comment (tonidero) [inline examples/purchase-tester/src/main/AndroidManifest.xml:18]: Had to do this to avoid the same activity from launching multiple times with the deep link... We might want to recommend developers to do the same.

File: examples/purchase-tester/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -15,12 +15,19 @@
...
                 android:exported="true">
             <intent-filter>
                 <action android:name="android.intent.action.MAIN" />
                 <category android:name="android.intent.category.LAUNCHER" />
             </intent-filter>
+            <intent-filter>
+                <action android:name="android.intent.action.VIEW" />
+                <category android:name="android.intent.category.BROWSABLE" />
+                <category android:name="android.intent.category.DEFAULT" />
+                <data android:scheme="revenuecatbilling" />
```
Comment (tonidero) [inline examples/purchase-tester/src/main/AndroidManifest.xml:29]: Currently hardcoded schema.

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt [hunk 3]
```diff
@@ -1,13 +1,30 @@
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        if (Purchases.isConfigured) {
+            Purchases.sharedInstance.handleDeepLink(intent)
+        }
     }
+
+    @OptIn(ExperimentalPreviewRevenueCatPurchasesAPI::class)
+    override fun onNewIntent(intent: Intent?) {
```
Comment (tonidero) [inline examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt:22]: Added this override since it might be needed if the activity is already there, like in this example

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt [hunk 4]
```diff
@@ -656,6 +662,57 @@ internal class Backend(
...
 
+    fun postRedeemRCBillingPurchase(
+        appUserID: String,
+        redemptionToken: String,
+        onSuccessHandler: () -> Unit,
+        onErrorHandler: (PurchasesError) -> Unit,
+    ) {
+        val endpoint = Endpoint.PostRedeemRCBillingPurchase(appUserID)
+        val path = endpoint.getPath()
+        val body = mapOf("new_app_user_id" to redemptionToken) // WIP: change this field to the actual field
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/common/Backend.kt:673]: We will create a new endpoint for this, instead of relying on the aliasing endpoint. For now, reusing that one and we need to use the corresponding parameter names.

File: purchases/src/defaults/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 5]
```diff
@@ -125,6 +126,15 @@ class Purchases internal constructor(
 
     // region Public Methods
 
+    /**
+     * Allows the SDK to handle the activities intent to process relevant deep links.
+     * @return A boolean indicating whether the SDK handled the deep link or ignored it.
+     */
+    @ExperimentalPreviewRevenueCatPurchasesAPI
+    fun handleDeepLink(intent: Intent): Boolean {
```
Comment (tonidero) [inline purchases/src/defaults/kotlin/com/revenuecat/purchases/Purchases.kt:134]: Note that this API means that the SDK needs to be configured BEFORE handling the deep link. I think we would need to add a wrapper around this to handle cases where the SDK is not configured.

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt [hunk 6]
```diff
@@ -237,6 +247,19 @@ internal class PurchasesOrchestrator(
...
+    fun handleDeepLink(intent: Intent): Boolean {
+        val deepLink = intent.data?.let { deepLinkHandler.parseDeepLink(it) } ?: return false
+
+        when (deepLink) {
+            is DeepLinkHandler.DeepLink.RedeemRCBPurchase -> {
+                rcBillingPurchaseRedemptionHelper.handleRedeemRCBPurchase(deepLink)
+            }
+        }
+        // WIP: This currently returns before the operation is completed. We might want to provide a way for devs
+        // to detect when the operation is started/completed.
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt:259]: We can probably take inspiration from the promoted purchases API in iOS: https://revenuecat.github.io/purchases-ios-docs/5.6.0/documentation/revenuecat/purchasesdelegate/purchases(_:readyforpromotedproduct:purchase:).

This allows the developer to control whether to perform the redemption and to get a result back.

File: purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/RCBillingPurchaseRedemptionHelper.kt [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+internal class RCBillingPurchaseRedemptionHelper(
+    private val backend: Backend,
+    private val identityManager: IdentityManager,
+    private val customerInfoHelper: CustomerInfoHelper,
+) {
+    var redeemRCBillingPurchaseListener: RedeemRCBillingPurchaseListener? = null
+
+    fun handleRedeemRCBPurchase(deepLink: DeepLink.RedeemRCBPurchase) {
+        debugLog("Detected RCB purchase redemption. Asking callback to initiate redeem.")
+        redeemRCBillingPurchaseListener?.handleRCBillingPurchaseRedemption { resultListener ->
```
Comment (tonidero) [inline purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/RCBillingPurchaseRedemptionHelper.kt:25]: Right now, it's mandatory to set the listener in order to process the redemption link or it will be ignored.

File: api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemRCBillingPurchaseListenerAPI.java [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+import com.revenuecat.purchases.interfaces.RedeemRCBillingPurchaseListener;
+
+import kotlin.Unit;
+import kotlin.jvm.functions.Function1;
+
+@SuppressWarnings({"unused"})
+final class RedeemRCBillingPurchaseListenerAPI {
+    static void checkListener(RedeemRCBillingPurchaseListener listener,
+                              Function1<RedeemRCBillingPurchaseListener.ResultListener, Unit> startRedemption) {
```
Comment (tonidero) [inline api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemRCBillingPurchaseListenerAPI.java:11]: This API is probably not great in Java 😅... I could create another interface to be passed in as parameter to the `handleRCBillingPurchaseRedemption` instead... But felt like 3 interfaces was a lot... Wdyt?

**Comment by tonidero** [comment]
Hi @RevenueCat/coresdk! While this PR is not final (the endpoint is wrong among other things), I would appreciate a look, specially on the new public APIs. I left comments in the places where it's still WIP. Lmk what you think! Also happy to chat over Slack if you have questions.

File: api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemRCBillingPurchaseListenerAPI.java (hunk 8 above)
Comment (vegaro) [reply api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemRCBillingPurchaseListenerAPI.java:15]: Hmmm yeah, not a fan, not only becuase of the `Function1`, but because of `Unit`

Comment (tonidero) [reply api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemRCBillingPurchaseListenerAPI.java]: Improved in https://github.com/RevenueCat/purchases-android/pull/1880/commits/2a16fffc73c1a805ce34a957295cf828ba4024c3. It adds another interface... But seemed like the best approach to make it look better. Lmk what you think!

**Review by JayShortway — COMMENTED** [review]
This is getting good! Had some thoughts, and a lot of questions to help me understand the design better haha, but we're well on our way!

File: purchases/src/main/kotlin/com/revenuecat/purchases/interfaces/RedeemRCBillingPurchaseListener.kt [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+ * Interface to handle the redemption of a RevenueCat Billing purchase.
+ */
+@ExperimentalPreviewRevenueCatPurchasesAPI
+interface RedeemRCBillingPurchaseListener {
+    /**
+     * Result of the redemption of a RevenueCat Billing purchase.
+     */
+    enum class RedeemResult {
+        SUCCESS,
+        ERROR,
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/interfaces/RedeemRCBillingPurchaseListener.kt:15]: I'm actually planning to make this a sealed class, to provide the CustomerInfo/PurchaseError. We also will want to add additional result cases like "Link expired" or something along those lines, so the developer can handle appropriately more easily, rather than parsing the `PurchasesError`.

[the remark this answers is not in the document]
Comment (JayShortway) [reply purchases/src/main/kotlin/com/revenuecat/purchases/interfaces/RedeemRCBillingPurchaseListener.kt:15]: That makes sense!

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt [hunk 10]
```diff
@@ -44,6 +47,25 @@ class MainApplication : Application(), UpdatedCustomerInfoListener {
         ).show()
         Log.d("CustomerInfoListener", "$message: $customerInfo")
     }
+
+    override fun handleRCBillingPurchaseRedemption(
+        redemptionStarter: RedeemRCBillingPurchaseListener.RedemptionStarter,
+    ) {
```
[the remark this answers is not in the document]
Comment (tonidero) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainApplication.kt:53]: > Their Activity's onCreate()/onNewIntent() I guess, depending on the launchMode?

Well, processing the deep link will be initiated from there, that's correct. However, to process the redemption link they can implement the `RedeemRCBillingPurchaseListener` interface any place they want. They just need to set the listener in the `Purchases.sharedInstance.redeemRCBillingPurchaseListener` at some point with that object. The current implementation caches the redemption token in memory until this listener is set.

> Is there any concern of a memory leak,

Well, depending on how long they hold the `redemptionStarter` there could be... This basically will hold some references to some internal classes of `Purchases`, but since most of those are singletons, it's mostly ok unless you reconfigure the SDK and keep an old reference to the `redemptionStarter`. I don't think we will be holding anything that would cause any big leaks on our side.

> What happens if the app never calls startRedemption()?

Nothing! We currently are not caching the links on disk so it won't survive restarts. We also won't try again from our side.

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt [hunk 11]
```diff
@@ -1,13 +1,25 @@
...
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        DeepLinkHandler.handleDeepLink(intent)
+    }
+
+    override fun onNewIntent(intent: Intent?) {
+        super.onNewIntent(intent)
+        if (intent != null) {
+            DeepLinkHandler.handleDeepLink(intent)
```
[the remark this answers is not in the document]
Comment (tonidero) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt:22]: Your understanding is correct! As for the reason, the developer needs to handle some cases like:
- If the token is expired, they probably want to display some kind of message indicating that a new link will be sent to their email
- If they want to hold the redemption to a certain point in the app's lifetime and maybe with a confirmation message.

Additionally, it will be easier to return a result for the redemption from the `startRedemption` than from the `handleDeepLink`, in case we support multiple types of deep links in the future.

If they actually don't want to handle any of those cases, they can just call `startRedemption()` immediately.

File: purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt [hunk 12]
```diff
@@ -45,7 +45,9 @@ internal class OfferingsManager(
                     LogIntent.DEBUG,
                     if (appInBackground) {
                         OfferingStrings.OFFERINGS_STALE_UPDATING_IN_BACKGROUND
-                    } else OfferingStrings.OFFERINGS_STALE_UPDATING_IN_FOREGROUND,
+                    } else {
+                        OfferingStrings.OFFERINGS_STALE_UPDATING_IN_FOREGROUND
+                    },
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/common/offerings/OfferingsManager.kt:50]: Ohh right sorry about that. I believe detekt is doing these for some reason... Will try to revert them before moving forward with these 👍

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt [hunk 13]
```diff
@@ -1218,6 +1248,22 @@ internal class PurchasesOrchestrator(
...
+        val unprocessedDeepLinks = mutableSetOf<DeepLinkParser.DeepLink>()
+        while (DeepLinkHandler.cachedLinks.isNotEmpty()) {
+            val deepLink = DeepLinkHandler.cachedLinks.first()
+            DeepLinkHandler.cachedLinks.remove(deepLink)
+            val processed = handleDeepLink(deepLink)
+            if (!processed) {
+                unprocessedDeepLinks.add(deepLink)
+            }
+        }
+        DeepLinkHandler.cachedLinks.addAll(unprocessedDeepLinks)
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt:1264]: Good call! Yeah I wanted to revisit this 👍

File: purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkHandler.kt [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+        @get:Synchronized
+        internal val cachedLinks = mutableSetOf<DeepLinkParser.DeepLink>()
+
+        /**
+         * Allows the RevenueCat SDK to handle the activity intent to process relevant deep links.
+         * @param intent The intent to process deep links for. Usually obtained from the Activity.
+         * @param shouldCache Whether the deep link should be cached for later processing if the SDK is not configured.
+         */
+        @Suppress("ReturnCount")
+        fun handleDeepLink(intent: Intent, shouldCache: Boolean = true): HandleResult {
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkHandler.kt:45]: Hmm honestly this was a bit of my preference to try to not bloat the `Purchases` class even more. Specially considering this can actually work without configuring the SDK... But honestly not a strong preference, if you prefer it to be inside `Purchases`, I'm ok moving it. As you said, it keeps everything into a single entry point.

File: purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkParser.kt [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+
+    @Suppress("ReturnCount")
+    fun parseDeepLink(data: Uri): DeepLink? {
+        if (data.host == REDEEM_RCB_PURCHASE_HOST) {
+            val redemptionToken = data.getQueryParameter("redemption_token")
+            if (redemptionToken.isNullOrBlank()) {
+                debugLog("Redemption token is missing RCBilling redemption deep link. Ignoring.")
+                return null
+            }
+            return DeepLink.RedeemRCBPurchase(redemptionToken)
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkParser.kt:23]: Yup that would be good, but for this PoC, we are actually using not final tokens (just some user ids actually). But yeah, good to take that into account 👍

File: purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/RCBillingPurchaseRedemptionHelper.kt [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+            }
+        })
+        return true
+    }
+
+    private fun handleResult(
+        resultListener: RedeemRCBillingPurchaseListener.ResultListener,
+        result: RedeemRCBillingPurchaseListener.RedeemResult,
+    ) {
+        dispatch { resultListener.handleResult(result) }
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/RCBillingPurchaseRedemptionHelper.kt:76]: Well, the result from the `Backend` operation will always be in our background dispatcher which is used internally. Seemed like a bad option to use that, so thought to just return in the main thread by default for now... We could also try to return in the same thread as the `startRedemption` was called, but that seemed more work.

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt]: So the thing here is that I want to achieve a few things:
- Iterate over the cachedLinks but keep them there if they are not processed immediately (which may happen if the redeem web purchase listener is not set when configuring the SDK).
- Clear them once they've been processed, even if not going through the redemption. 

So I'm thinking of a different approach of still using a set but wrapping the API around the `forEachCachedLink` API you suggested. I just pushed those changes. Lmk what you think!

File: purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkHandler.kt [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+        /**
+         * Deep link is not recognized by RevenueCat's SDK so it's ignored.
+         */
+        IGNORED,
+    }
+
+    companion object {
+        @get:Synchronized
+        internal val cachedLinks = mutableSetOf<DeepLinkParser.DeepLink>()
```
[the remark this answers is not in the document]
Comment (tonidero) [reply purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkHandler.kt]: Changed to use a "synchronized set" instead... Basically want to avoid someone clicking on the link twice and adding it twice to the queue... As for handling in order, I don't believe it's a big deal even if processed out of order at least for now. Lmk if you think otherwise!

File: purchases/src/defaults/kotlin/com/revenuecat/purchases/Purchases.kt [hunk 18]
```diff
@@ -99,6 +101,18 @@ class Purchases internal constructor(
             purchasesOrchestrator.updatedCustomerInfoListener = value
         }
 
+    /**
+     * Set a value to be able to handle redemptions of RevenueCat web purchases.
+     * You can set to null to destroy the listener.
+     */
+    @ExperimentalPreviewRevenueCatPurchasesAPI
+    var redeemWebPurchaseListener: RedeemWebPurchaseListener?
```
Comment (tonidero) [inline purchases/src/defaults/kotlin/com/revenuecat/purchases/Purchases.kt:109]: I wondered about making this a parameter for the SDK configuration… but since this will need to be something that handles the UI for redeeming web purchases, it seemed more flexible to be able to set it later, but lmk what you think!

File: api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemWebPurchaseListenerAPI.java [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+        webPurchaseRedeemer.redeemWebPurchase(resultListener);
+    }
+
+    static void checkResultListener(RedeemWebPurchaseListener.ResultListener resultListener,
+                                    RedeemWebPurchaseListener.Result result) {
+        resultListener.handleResult(result);
+    }
+
+    static void checkRedeemResult(RedeemWebPurchaseListener.Result result) {
+        if (result instanceof RedeemWebPurchaseListener.Result.Success) {
```
Comment (vegaro) [inline api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemWebPurchaseListenerAPI.java:29]: Maybe we could add a function `isSuccess()` to the sealed class that always returns true for `Success`? Just some syntactic sugar

Comment (tonidero) [reply api-tester/src/defaults/java/com/revenuecat/apitester/java/RedeemWebPurchaseListenerAPI.java:29]: Done in https://github.com/RevenueCat/purchases-android/pull/1880/commits/810da26702bcc09a9349ba59aff76afefccbe93c

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (JayShortway) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt:22]: I understand the scenarios we'd like to cover. I'm a bit concerned about the long chain of callbacks here. We have:
1. `onCreate()` / `onNewIntent()`, in which `handleDeepLink()` is called.
2. `handleWebPurchaseRedemption()`, in which `redeemWebPurchase()` is called.
3. `handleResult()`, containing the result of the redemption. 

It's hard to follow the flow of the code through all these callbacks, if a developer is trying to reason about their own web redemption code. It possibly involves going through multiple files to see what each of these callbacks is doing. 

I have 2 alternatives in mind, mostly intended to challenge your assumptions haha.

### Idea 1: no RedeemWebPurchaseListener

One idea I had is to remove `RedeemWebPurchaseListener.handleWebPurchaseRedemption()` from the equation. We could define an instance method, e.g. in `Purchases`, along the lines of: 

```kotlin
fun getWebPurchaseRedeemerOrNull(intent: Intent): RedeemWebPurchaseListener.WebPurchaseRedeemer?
```

This will return null if the intent is not a web purchase intent. This can then be used like so: 

```kotlin
override fun onNewIntent(intent: Intent) {
    Purchases.sharedInstance
        .getWebPurchaseRedeemerOrNull(intent)
        ?.redeemWebPurchase { result -> }
}
```

Devs can keep a hold of the `WebPurchaseRedeemer`, or the `Intent`, if they don't want to redeem the purchase right away. 

### Idea 2: no RedeemWebPurchaseListener & no WebPurchaseRedeemer

To simplify even further, we could remove the concept of a `WebPurchaseRedeemer`. We could define an instance method in `Purchases` along the lines of: 

```kotlin
fun redeemWebPurchase(intent: Intent, resultListener: ResultListener)
```

The `ResultListener` would return an error result if the intent is not a web purchase intent. (If devs want to check whether the intent is a web purchase intent before calling `redeemWebPurchase()`, we can add a method `isWebPurchaseRedemptionIntent()` that they can use.)

This can then be used like so: 

```kotlin
override fun onNewIntent(intent: Intent) {
    Purchases.sharedInstance.redeemWebPurchase(intent) { result -> }
}
```

In this case, if devs don't want to redeem right away, they can keep a hold of the `Intent`, and call `redeemWebPurchase()` whenever they want.

-------
  
  
Again, these are mostly intended to challenge your assumptions and see if there are opportunities to simplify. If we can get away with it, I like idea 2 best because it's by far the simplest and removes a lot of concepts for devs to understand.

File: purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (JayShortway) [reply purchases/src/main/kotlin/com/revenuecat/purchases/PurchasesOrchestrator.kt]: Very nice approach! 🙌

File: purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkHandler.kt (hunk 17 above)
[the remark this answers is not in the document]
Comment (JayShortway) [reply purchases/src/main/kotlin/com/revenuecat/purchases/deeplinks/DeepLinkHandler.kt]: I think that works! I'm not sure if handling out of order is a problem. 🤔 If we think it might be, we could look into using a `LinkedHashSet`. Otherwise I think this is fine.

**Review by JayShortway — COMMENTED** [review]
Very nice! Basically 2 open points left for me: 
- Some ideas to simplify the API.
- Maybe we can require the SDK to be configured, for now.

Happy to have a broader discussion about this if you prefer!

File: examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (tonidero) [reply examples/purchase-tester/src/main/java/com/revenuecat/purchasetester/MainActivity.kt:22]: Those are very valid thoughts! So the main points with both of those approaches are IMO:
- The developer doesn't have a way to detect if there is a redemption link beforehand (unless they check the intent themselves which wouldn't be a great API IMO). This could be useful if the developer wants to display some UI like: "You are redeeming your web Purchase!" "Yes/No". Maybe not a needed use case for now, but just thinking about this case.
- These work better if we don't support pre-configuration caching, which as you said separately, we may consider not supporting for now.

Let's chat about this over Slack/in person 👍

**Comment by tonidero** [comment]
I'm closing this PR since looks like we will end up going with the approach in #1889
